File: ts/S2loxClass.ts
import Interpreter from "Interpreter";
import S2loxCallable from "S2loxCallable";
import S2loxFunction from "S2loxFunction";
import S2loxInstance from "S2loxInstance";
import S2ltype from "S2ltype";

export default class S2loxClass implements S2loxCallable{
    constructor(
	public readonly name: string,
	public readonly superClass: S2loxClass | null,
	private readonly methods: Map<string, S2loxFunction>,
    ) {}

    public findMethod(name: string): S2loxFunction | null{
	if(this.methods.has(name)){
	    let method = this.methods.get(name);
	    if(method === undefined) return null;
	    return method;
	}

	if(this.superClass != null){
	    return this.superClass.findMethod(name);
	}

	return null;
    }

    public toString(): string{
	return `<class ${this.name}>`;
    }

    public arity(): number{
	let initializer = this.findMethod("init");
	if(initializer == null) return 0;
	return initializer.arity();
    }

    public call(interpreter: Interpreter, args: S2ltype[]): S2ltype{
	let instance = new S2loxInstance(this);
	let initializer = this.findMethod("init");
	if(initializer !== null){
	    initializer.bind(instance).call(interpreter, args);
	}
	return instance;
    }
}

File: java/com/thmtvz/s2lox/S2loxClass.java
package com.thmtvz.s2lox;

import java.util.List;
import java.util.Map;

class S2loxClass implements S2loxCallable {
    
    final String name;
    final S2loxClass superClass; 
    private final Map<String, S2loxFunction> methods;
    
    S2loxClass(String name, S2loxClass superClass, Map<String, S2loxFunction> methods){
	this.superClass = superClass;
	this.name = name;
	this.methods = methods;
    }

    S2loxFunction findMethod(String name){
	if(methods.containsKey(name)){
	    return methods.get(name);
	}

	if(superClass != null){
	    return superClass.findMethod(name);
	}

	return null;
    }

    @Override
    public String toString(){
	return "<class " + this.name + ">";
    }
    
    @Override
    public int arity(){
	S2loxFunction initializer = findMethod("init");
	if(initializer == null) return 0;
	return initializer.arity();
    }
    
    @Override
    public Object call(Interpreter interpreter, List<Object> arguments){
	S2loxInstance instance = new S2loxInstance(this);
	S2loxFunction initializer = findMethod("init");
	if(initializer != null){
	    initializer.bind(instance).call(interpreter, arguments);
	}
	return instance;
    }
}
